src/OpenAIGym.jl

```diff
@@ -19,40 +20,47 @@ const _py_envs = Dict{String,Any}()
 abstract type AbstractGymEnv <: AbstractEnvironment end
 
 "A simple wrapper around the OpenAI gym environments to add to the Reinforce framework"
-mutable struct GymEnv <: AbstractGymEnv
+mutable struct GymEnv{T} <: AbstractGymEnv
     name::String
-    pyenv  # the python "env" object
-    state
+    pyenv::PyObject   # the python "env" object
+    pystep::PyObject  # the python env.step function
+    pyreset::PyObject # the python env.reset function
+    pystate::PyObject # the state array object referenced by the PyArray state.o
+    pystepres::PyObject # used to make stepping the env slightly more efficient
+    info::PyObject    # store it as a PyObject for speed, since often unused
+    state::T
     reward::Float64
     total_reward::Float64
     actions::AbstractSet
     done::Bool
-    info::Dict
-    GymEnv(name,pyenv) = new(name,pyenv)
-end
-GymEnv(name) = gym(name)
-
-function Reinforce.reset!(env::GymEnv)
-    env.state = env.pyenv[:reset]()
-    env.reward = 0.0
-    env.total_reward = 0.0
-    env.actions = actions(env, nothing)
-    env.done = false
+    function GymEnv{T}(name, pyenv, pystate, state) where T
+        env = new{T}(name, pyenv, pyenv["step"], pyenv["reset"],
+                                 pystate, PyNULL(), PyNULL(), state)
+        reset!(env)
+        env
+    end
 end
 
-function gym(name::AbstractString)
+function GymEnv(name; stateT=PyArray)
     env = if name in ("Soccer-v0", "SoccerEmptyGoal-v0")
         Base.copy!(gym_soccer, pyimport("gym_soccer"))
         get!(_py_envs, name) do
-            GymEnv(name, pygym[:make](name))
+            GymEnv(name, pygym[:make](name), stateT)
         end
     else
-        GymEnv(name, pygym[:make](name))
+        GymEnv(name, pygym[:make](name), stateT)
     end
     reset!(env)
     env
 end
 
+function GymEnv(name, pyenv, stateT)
+    pystate = pycall(pyenv["reset"], PyObject)
```

Comment:
`pyenv[:reset]`

Reply:
pyenv["reset"] is slightly faster since it doesn't try to convert pyenv["reset"] to a Julia type, and I want the result to be a PyObject, so I use the full pycall syntax (https://github.com/JuliaPy/PyCall.jl#pyobject) This will change to pyenv.reset, soon-ish anyway when get/setproperty gets added to PyCall